doc/sphinx/proof-engine/tactics.rst

```diff
@@ -898,211 +1109,156 @@ Applying theorems
+      .. coqtop: all
 
-      This is equivalent to the composed tactic :n:`intro @ident; ... ; intro @ident`.
+         intros m n H x.
 
-   .. tacv:: intros until @ident
+.. tacn:: intros {* @intropattern }
+          intros until {| @ident | @natural }
 
-      This repeats intro until it meets a premise of the goal having the
-      form :n:`(@ident : @type)` and discharges the variable
-      named :token:`ident` of the current goal.
+      The first form introduces zero or more items into the context from the goal
+      using the :n:`intro` tactic.
+      If :n:`@intropattern` is not specified, the tactic repeats :tacn:`intro`
+      until it reaches the :term:`head constant`.  The :n:`@intropattern` is
+      often just a list of :n:`ident`\s, but other forms can also be specified
+      in order to, for example, introduce all quantified variables (`*`);
+      introduce all hypotheses before `->` and all quantified variables (`**`);
+      split terms such as `A /\\ B` and pick a fresh name with a given prefix (`?X`).
+      See See :ref:`intropatterns`.
 
-      .. exn:: No such hypothesis in current goal.
-         :undocumented:
+      This form never reduces the :term:`head constant` and it never fails.
+      It can leave the proof context unchanged.
 
-   .. tacv:: intros until @natural
+      The second form repeats :n:`intro` until after it has introduced a dependent
+      premise with the name :n:`@ident` or has introduced :n:`@natural`
+      non-dependent premises.
 
-      This repeats :tacn:`intro` until the :token:`natural`\-th non-dependent
-      product.
+      .. todo PR: should we discourage use of "until" because it lets Coq pick the names?
+         (maintainability issue)
+         Also, shall I make "intros until" a separate command?
 
       .. example::
 
-         On the subgoal :g:`forall x y : nat, x = y -> y = x` the
-         tactic :n:`intros until 1` is equivalent to :n:`intros x y H`,
-         as :g:`x = y -> y = x` is the first non-dependent product.
+         On the goal :g:`forall x y : nat, x = y -> y = x` the
+         tactic :n:`intros until 1` is equivalent to :n:`intros x y H`
+         because :g:`x = y -> y = x` is the first non-dependent premise.
+
+         .. todo PR: think I should drop the following example.  z is not
+            used on the RHS--unless we can change it so its useful
 
-         On the subgoal :g:`forall x y z : nat, x = y -> y = x` the
+         On the goal :g:`forall x y z : nat, x = y -> y = x` the
          tactic :n:`intros until 1` is equivalent to :n:`intros x y z`
          as the product on :g:`z` can be rewritten as a non-dependent
-         product: :g:`forall x y : nat, nat -> x = y -> y = x`.
+         premise: :g:`forall x y : nat, nat -> x = y -> y = x`.
 
       .. exn:: No such hypothesis in current goal.
 
          This happens when :token:`natural` is 0 or is greater than the number of
          non-dependent products of the goal.
```

Comment:
As I noted above, the message no longer exists.  I don't think it's urgent to add these 2 messages, either.

Reply:
In this case, these error messages are not generic but specific to this tactic. I don't know why you are talking of the (non-)urgency of documenting error messages, but if you want me to stop digging into the behavior of tactics while we edit this chapter, I can stop.

Reply:
I'll add these, but I think it's quite hard to do a good job documenting error messages, particularly if it's not done systematically as a project.  I think the first step is to create tooling to somehow extract all the error messages from the code, which makes it possible to strategize, to prioritize improvements and to track and review ongoing changes to messages.  Associating messages with specific tactics likely requires much more time reading source code than it has for getting the syntax well-documented.  Also unclear where messages used by multiple tactics should appear.  Looking in the messages index is a good strategy for users.

Another part of it is that we should make messages as clear as possible so users generally don't need to reach for the doc to understand them, though certainly need to be written up.

Sure, keep digging into tactics behavior but no deeper than six feet.

Reply:
This example is wrong:

![image](https://user-images.githubusercontent.com/1253341/188238965-a5ebe5e4-fbc2-4819-91e4-793c2aa2fb58.png)

Reply:
Also not keen about using the terms "dependent premise" and "non-dependent premise" anywhere.  Much more understandable to simply show the form (forall or ->).

Reply:
Yes, error messages would deserve a project on their own, although I'm unfortunately of the opinion that we cannot instrument the Coq code to get the list of all error messages without significantly refactoring the way error messages are generated (which is probably a good idea anyway). Without such a refactoring, a documenting effort around error messages could be to systematically introduce tests for all the error messages that we can find and for every tactic that we can find that triggers them. At least that would help have a good cartography, know what to document and what to improve, and help with making sure that documentation is updated when error messages change.

> What's the significance of "**quantified** hypothesis?

I think it should be understood as meaning strictly the same as a "quantified variable" or "dependent premise". There seem to be too many names for exactly the same thing :/

> Also, I don't see the point of the deprecation if `until 1` is still supported:
> 
> `"intros until 0" is deprecated, use "intros *"; instead of "induction 0" and "destruct 0" use explicitly a name." [deprecated-intros-until-0,tactics]`

Yeah, I agree that deprecating is probably not strictly useful. I guess the point is mostly to remove redundant features.

>This example is wrong:

Yeah, and your fix is right.

>Much more understandable to simply show the form (forall or ->).

I disagree. I think having a consistent terminology helps brevity which can help clarity (especially when explaining more complex behaviors). That being said "non-dependent premise" is a mouthful, and I suggested somewhere else, there are many places where it can be shortened to just "premise". As for "dependent premise", it can also be called "quantified variable", which may be less jargon-esque.

Reply:
>>Much more understandable to simply show the form (forall or ->).

> I disagree. I think having a consistent terminology helps brevity which can help clarity (especially when explaining more complex behaviors).

Showing the form is like showing an example, frequently much easier to grasp than a verbal description.  Perhaps not right for all cases.

Using "premise" to generally mean "non-dependent premise" works for me.  It is essentially the common usage.  OTOH, "dependent premise" is not commonly used (e.g. Google search for a definition doesn't match Coq's usage).

WDYT about generally referring to "dependent premises" as "foralls" or "forall variables"?  While no one would use that term  in an academic paper, I think it's instantly understandable even without explanation.  Though of course somewhere we can give the standard terminology.  (Or what other term would you suggest?  "universally quantified variable"? yuck)

Reply:
I think it's pretty standard to refer to them simply as "variables". "exists variables" are typically termed constants or definitions.

Reply:
>"exists variables" are typically termed constants or definitions.

Really?

"forall variables" sounds reasonable to me, but I wouldn't have an issue with using "universally quantified variables" here and there and "variables" for short in other places.

Reply:
"dependent premise" and "dependent product" are synonyms, right?  Would it make sense to use one consistently (premise) and mention the other in a single place as a synonym?

I see a number of uses of "dependent", such as "dependent pattern matching", "non-dependent binders", "dependent subgoals", "Non-dependent implicit arguments", "dependent evars", "non-dependent construct", "dependent equality", "non-dependent elimination", "dependent types", "dependent hypotheses" and others.

Perhaps it would be good to define "dependent" and "non-dependent" as concepts in addition to "dependent premise", etc.  It means that two or some group of items don't share any variables, right?  Somehow I haven't given any thought to this before...

Reply:
FWIW, "nondependent" is preferred in the US while "non-dependent" is preferred in the UK.  Mostly we use the latter in the doc and I think it's a bit easier to read that way.

Reply:
>"dependent premise" and "dependent product" are synonyms, right?

IINM "dependent product" is the full term built with a forall. So `forall (x : A), @term`, whereas "dependent premise" is the `(x : A)` part.

>I see a number of uses of "dependent", such as "dependent pattern matching", "non-dependent binders", "dependent subgoals", "Non-dependent implicit arguments", "dependent evars", "non-dependent construct", "dependent equality", "non-dependent elimination", "dependent types", "dependent hypotheses" and others.

Other examples include "dependent pair", "dependent record"... Usually, "dependent" means that the variable defined in a first part of the construct may be used in the type of the latter part of the construct.

>Perhaps it would be good to define "dependent" and "non-dependent" as concepts

Sounds good. This would make sense to define these concepts in some early place in the manual, such as https://coq.inria.fr/refman/language/core/basic.html#essential-vocabulary. But better to keep that for a separate PR where we can involve others in reviewing / advising.